strategy_infura.go

```diff
@@ -0,0 +1,443 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/autoscaling"
+	"github.com/aws/aws-sdk-go/service/elbv2"
+)
+
+func strategyInfuraRefresh(ctx context.Context, region string, asgName string) error {
+	// add a timeout
+	ctx, cancel := context.WithTimeout(ctx, time.Hour)
+	defer cancel()
+
+	sess := session.Must(session.NewSession(aws.NewConfig().WithRegion(region)))
+
+	start := time.Now()
+
+	originalAsg, err := getASGDetails(ctx, sess, asgName)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Refresh (strategy: infura-refresh) started on ASG %s\n", asgName)
+	fmt.Printf("Existing instances:\n")
+	for _, inst := range originalAsg.instances {
+		fmt.Printf("\t%s\n", inst)
+	}
+
+	// Record rollbacks during the procedure to get back on a reasonable state if the context get cancelled
+	// var rollbacks []func(ctx context.Context) error
+	//
+	// rollbacks = append(rollbacks, func(ctx context.Context) error {
+	// 	return setASGCapacities(ctx, as, asgName, originalAsg.maxSize, originalAsg.desiredCapacity)
+	// })
+
+	// Set instance protection on the old instances to be sure we keep having them around in case of failure
+	fmt.Printf("Enable instance protection on old instances\n")
+	err = setInstanceProtection(ctx, sess, asgName, originalAsg.instances)
+	if err != nil {
+		return err
+	}
+
+	// Double the ASG sizes
+	fmt.Printf("Double the ASG size:\n\tmaxSize %d-->%d\n\tdesiredCapacity %d-->%d\n",
+		originalAsg.maxSize, 2*originalAsg.maxSize, originalAsg.desiredCapacity, 2*originalAsg.desiredCapacity)
+	err = setASGCapacities(ctx, sess, asgName, 2*originalAsg.maxSize, 2*originalAsg.desiredCapacity)
+	if err != nil {
+		return err
+	}
+
+	fmt.Printf("Detect new instances and wait for them to be healthy\n")
+	newInstances, err := waitForNewInstances(ctx, sess, asgName, int(originalAsg.desiredCapacity), originalAsg.tg, originalAsg.instances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// protect the new instances
+	fmt.Printf("Enable instance protection on the new instances\n")
+	err = setInstanceProtection(ctx, sess, asgName, newInstances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// unprotect the old instances
+	fmt.Printf("Remove instance protection on the old instances\n")
+	err = removeInstanceProtection(ctx, sess, asgName, originalAsg.instances)
+	if err != nil {
+		// rollback !
+		// TODO
+	}
+
+	// get back to normal size, cull the old instances
+	fmt.Printf("Set back the ASG size:\n\tmaxSize %d-->%d\n\tdesiredCapacity %d-->%d\n",
+		2*originalAsg.maxSize, originalAsg.maxSize, 2*originalAsg.desiredCapacity, originalAsg.desiredCapacity)
+	err = setASGCapacities(ctx, sess, asgName, originalAsg.maxSize, originalAsg.desiredCapacity)
```

Comment:
This would be best to do before the scale in protection removal step above, or the ASG may start an AZ rebalancing process and start different instances to replace those that are no longer protected (rebalancing is paused permanently on most hammerhead ASGs, but not guaranteed to be that way for others).

Reply:
I believe this is not a problem anymore with the new structure. Instances can be terminated,replaced ... during the refresh. Healthy and ready instances will be protected until the process end.